civiform/civiform
# [Program Card Images] Add UI for admin to upload a program image.

### Description

Adds a UI that allows admins to upload a program image.

Breakdown of what's in this PR:
1. Frontend: Adds new "Choose program image" and "Save program image" buttons to the admin program images page.
         - Note: The "Save image" button is separate from the "Save image description" button because the file upload form needs to be separate from any other data -- see https://docs.google.com/document/d/1HU4gtZz8e4wgsR4xZ4tAkZlZd4bAd4fM8i5Ela9NoHI/edit#heading=h.qmw6qx4cx1a5.
         - `FileUploadViewStrategy` was the old class responsible for rendering all the HTML elements required for the file upload question on the applicant side. The code that's needed for both applicant file upload and admin image upload  remains in `FileUploadViewStrategy` & subclasses; the code that's only needed for the applicant file upload has moved to `ApplicantFileUploadRenderer`.
         - Uses `PublicStorageClient` defined in https://github.com/civiform/civiform/pull/6049 to create the `<form>` element correctly.
3. Backend: Once the file has been uploaded to cloud storage, saves a key identifying the file to the program definition.
4. Database: Add a new column to the `programs` table storing the file key.

Note that this UI is not at all final and is just to get all the pieces working together. We'll need to do another pass at the UI later to make it look much better :)

## Release notes

### Checklist

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)


#### User visible changes


#### New Features


### Instructions for manual testing

1. Edit your `/etc/hosts` file to also contain this line:
```
127.0.0.1 civiform-local-s3-public.s3.localhost.localstack.cloud
```

2. Enable the program card images flag.
3. As an admin, edit a program and then click the "Edit program image" button.
4. On that page, click "Choose program image" and upload an image.
5. Click "Save image".
6. Verify the page reloads and displays the image that was uploaded.
7. Leave that page then return to the program image page and verify that the previously uploaded image is displayed.

### Issue(s) this completes

Epic: #5676 

Fixes #5999
